### ISSUE TITLE
Controller keeps crashing

### ISSUE THREAD

**Issue by uxth** [argoproj/argo-workflows#5808]
## Summary 

Argo controller keeps crashing

## Diagnostics

What Kubernetes provider are you using? 
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:31:21Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.0", GitCommit:"cb303e613a121a29364f75cc67d3d580833a7479", GitTreeState:"clean", BuildDate:"2021-04-08T16:25:06Z", GoVersion:"go1.16.1", Compiler:"gc", Platform:"linux/amd64"}

What version of Argo Workflows are you running? 
argo: v2.9.1
  BuildDate: 2020-07-03T15:06:40Z
  GitCommit: 6b967d08c0a142aaa278538f2407c28de467262e
  GitTreeState: clean
  GitTag: v2.9.1
  GoVersion: go1.13.4
  Compiler: gc
  Platform: linux/amd64

log file attached

[logs.txt](https://github.com/argoproj/argo-workflows/files/6419283/logs.txt)

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
```
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).getOutboundNodes(0xc0004e5ad0, 0xc0006de720, 0x1b, 0x2e4f448, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2274 +0x1a5 fp=0xc020b8d278 sp=0xc020b8d058 pc=0x1a6a6c5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).getOutboundNodes(0xc0004e5ad0, 0xc0006de440, 0x1b, 0x2e4f448, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2274 +0x1a5 fp=0xc020b8d498 sp=0xc020b8d278 pc=0x1a6a6c5
github.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).getOutboundNodes(0xc0004e5ad0, 0xc0006de800, 0x1c, 0x2e4f448, 0x0, 0x0)
	/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2274 +0x1a5 fp=0xc020b8d6b8 sp=0xc020b8d498 pc=0x1a6a6c5
...additional frames elided...
created by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).startLeading
```

This suggests that there is something wrong with one of your workflows. Maybe a DAG, where there is a cycle. This results in a stack-overflow and then crash.

Can you please determine which workflow is the problem and attached it to this ticket?

**Comment by uxth** [comment]
In fact I ran the dag diamond example from this repo.

What I guess is not normal use is:

I created 3 namespaces, argo00, argo01, workflow
I installed namespaced argo instance in argo00 and argo01, they both have the workflow as managed namespace. And also I have change the deployment replicas to 2.
then I submit the dag diamond example,  but from the workflow log I can see the workflow succeeded.

but I am not sure why the controller thinks there is cycles.

BTW, I have another question about HA.
in this case I set the deployment replicas to 2, and from the argo server service and controller service, I can see both of the instances showed up in the endpoints. Does this mean the 2 replicas are working together to share the traffic?
From the official doc, it said controller could only have one instance working at one time, set replicas to 2 so that one of them is hot stand-by.
I am confused here.
